
<h1><center> Homework 3 - Pandas Exercises </h1></center>

## Introduction

There are two parts in this homework assignment. In Part I, you will use the "student alcohol consumption" dataset, and you will be using the 'apply' method from Pandas DataFrame class. In Part II, you will work with the 'automobile' datasets, with which you have to clean the 'NaN's and apply other statistical analysis methods.

Follow along with the steps in each part and fulfill the tasks. 

+ Review the Jupyter notebooks from the lecture before getting started.
+ Change the name of your notebook to 'hw3_LastnameFirstInitial.ipynb' and submit it to Canvas. 

## Part I - Student Alcohol Consumption (50 points in total, 5 points for each step)

### Step 1. Import the necessary libraries, e.g. Pandas and Numpy

In [8]:
import numpy as np
import pandas as pd



### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/BlueJayADAL/DS200/master/datasets/student-alcohol.csv). Simply click on this link and copy the URL from your browser. Assign the URL to a variable called csv_url, then use the URL to create a DataFrame named df.

In [11]:
url = 'https://raw.githubusercontent.com/BlueJayADAL/DS200/master/datasets/student-alcohol.csv'




In [12]:
df = pd.read_csv(url)

### Step 3. Show the first 5 rows from your imported dataset and show the date types of each column. 

In [13]:
df.head()



school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

In [14]:
df.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

### Step 4. For the purpose of this exercise, slice the DataFrame 'df' from the column 'school' until the column 'guardian', then show the first 5 rows of the sliced DataFrame.
* Hint: you can use the 'loc' methods with slicing ':' to grab a collection of adjacent collumns from a DataFrame. Check the documentation of 'loc'.

In [15]:
stud_alcoh = df.loc[:,'school':'guardian']



stud_alcoh.head()



school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  
0  course   mother  
1  course   father  
2   other   mother  
3    home   mother  
4    home   father

### Step 5. Write a lambda expression that covert an input string into all capital letters. Assign the lambda to a variable named `captalizer`

In [16]:
# Type down your lambda expression here, and give it a name that you can refer to.
captalizer = (lambda cap: cap.upper())

### Step 6. Note that the strings in columns 'Mjob' and 'Fjob' are in lower cases. Use the lambda expression from Step 5 to capitalize both Mjob and Fjob columns, and show the first 5 rows from your DataFrame.
* Hint: use the apply method to apply the lambda expression to each one of your column values. 
* Hint: if you don't see the change of the columns from the DataFrame after you use the apply method, that means you haven't modified the orginal DataFrame. 

In [29]:
stud_alcoh['Mjob'] = stud_alcoh['Mjob'].apply(captalizer)

stud_alcoh['Fjob'] = stud_alcoh['Fjob'].apply(captalizer)

stud_alcoh.head()


school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  AT_HOME   TEACHER   
1     GP   F   17       U     GT3       T     1     1  AT_HOME     OTHER   
2     GP   F   15       U     LE3       T     1     1  AT_HOME     OTHER   
3     GP   F   15       U     GT3       T     4     2   HEALTH  SERVICES   
4     GP   F   16       U     GT3       T     3     3    OTHER     OTHER   

   reason guardian  
0  course   mother  
1  course   father  
2   other   mother  
3    home   mother  
4    home   father

### Step 7. Grab the last 5 rows from the DataFrame.

In [30]:
stud_alcoh.tail()




school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
390     MS   M   20       U     LE3       A     2     2  SERVICES  SERVICES   
391     MS   M   17       U     LE3       T     3     1  SERVICES  SERVICES   
392     MS   M   21       R     GT3       T     1     1     OTHER     OTHER   
393     MS   M   18       R     LE3       T     3     2  SERVICES     OTHER   
394     MS   M   19       U     LE3       T     1     1     OTHER   AT_HOME   

     reason guardian  
390  course    other  
391  course   mother  
392  course    other  
393  course   mother  
394  course   father

### Step 8. Suppose you are in a country that people can legally start to drink alcohol when they are older than 16 (not including 16). Create a function called `check_drinker` that returns a boolean value indicating whether the drinker is legal to drink. Then create a new column for the DataFrame called `legal_drinker` by applying the `check_drinker` function to every students on the DataFrame. Show the first 5 rows of the resulting DataFrame.

In [32]:
# Create the drinker function here
def check_drinker(x):
    if x > 16:
        return True
    else:
        return False




In [33]:
stud_alcoh['legal_drinker'] = stud_alcoh['age'].apply(check_drinker)

stud_alcoh.head()


school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  AT_HOME   TEACHER   
1     GP   F   17       U     GT3       T     1     1  AT_HOME     OTHER   
2     GP   F   15       U     LE3       T     1     1  AT_HOME     OTHER   
3     GP   F   15       U     GT3       T     4     2   HEALTH  SERVICES   
4     GP   F   16       U     GT3       T     3     3    OTHER     OTHER   

   reason guardian  legal_drinker  
0  course   mother           True  
1  course   father           True  
2   other   mother          False  
3    home   mother          False  
4    home   father          False

### Step 9. Multiply every numeric column from the dataset by 10. Don't change it permanently. 
##### I know this makes no sense, don't forget it is just an exercise
* Hint: carefully design your times10 function so that it only works on the numerical columns (the other data types won't be affected)
* Hint: note that the 'apply' may not be working here because 'apply' takes a Series or Series from a DataFrame as the input to be applied. However, here we want to apply a function to a DataFrame elementwise. 

In [35]:
# It's a good idea to check the types of different columns first
stud_alcoh.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   school         395 non-null    object
 1   sex            395 non-null    object
 2   age            395 non-null    int64 
 3   address        395 non-null    object
 4   famsize        395 non-null    object
 5   Pstatus        395 non-null    object
 6   Medu           395 non-null    int64 
 7   Fedu           395 non-null    int64 
 8   Mjob           395 non-null    object
 9   Fjob           395 non-null    object
 10  reason         395 non-null    object
 11  guardian       395 non-null    object
 12  legal_drinker  395 non-null    bool  
dtypes: bool(1), int64(3), object(9)
memory usage: 37.5+ KB


In [25]:
# Define the times10 function here
def times10(x):
    if type(x) == type(1):
        return x * 10
    else:
        return x




In [36]:
# Multiply the numerical columns by 10 and show the first 10 rows here
stud_alcoh.applymap(times10)




school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F  180       U     GT3       A    40    40   AT_HOME   TEACHER   
1       GP   F  170       U     GT3       T    10    10   AT_HOME     OTHER   
2       GP   F  150       U     LE3       T    10    10   AT_HOME     OTHER   
3       GP   F  150       U     GT3       T    40    20    HEALTH  SERVICES   
4       GP   F  160       U     GT3       T    30    30     OTHER     OTHER   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M  200       U     LE3       A    20    20  SERVICES  SERVICES   
391     MS   M  170       U     LE3       T    30    10  SERVICES  SERVICES   
392     MS   M  210       R     GT3       T    10    10     OTHER     OTHER   
393     MS   M  180       R     LE3       T    30    20  SERVICES     OTHER   
394     MS   M  190       U     LE3       T    10    10     OTHER   AT_HOME   

     reason guardian  legal_drinker  
0    course   mother           True  
1    course   father           True  
2     other   mother          False  
3      home   mother          False  
4      home   father          False  
..      ...      ...            ...  
390  course    other           True  
391  course   mother           True  
392  course    other           True  
393  course   mother           True  
394  course   father           True  

[395 rows x 13 columns]

### Step 10. Check for each type of family size (`famsize`), what are the students average ages. And check for each type of parental status (`Pstatus`), what is the mother's longest education years (`Medu`)?

In [37]:
stud_alcoh.groupby(by = 'famsize')['age'].mean()




famsize
GT3    16.66548
LE3    16.77193
Name: age, dtype: float64

In [38]:
stud_alcoh.groupby(by = 'Pstatus')['Medu'].max()




Pstatus
A    4
T    4
Name: Medu, dtype: int64

____

## Part II - The Automobile Dataset (50 points in total; 5 points for each step)

### Introduction:

The following exercise utilizes data from [UC Irvine Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/Auto+MPG)

### Step 1. Import the necessary libraries. E.g. Pandas and Numpy

In [39]:
import numpy as np
import pandas as pd

### Step 2. Import the two datasets [cars1](https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/05_Merge/Auto_MPG/cars1.csv) and [cars2](https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/05_Merge/Auto_MPG/cars2.csv). Assign each to a DataFrame variable called `cars1` and `cars2` separately.

In [40]:
cars1 = pd.read_csv('https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/05_Merge/Auto_MPG/cars1.csv')


cars2 = pd.read_csv('https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/05_Merge/Auto_MPG/cars2.csv')



### Step 3. Show the first 5 rows of each DataFrame you got from step 2. And show all the column data types.

In [41]:
cars1.head()



mpg  cylinders  displacement horsepower  weight  acceleration  model  \
0  18.0          8           307        130    3504          12.0     70   
1  15.0          8           350        165    3693          11.5     70   
2  18.0          8           318        150    3436          11.0     70   
3  16.0          8           304        150    3433          12.0     70   
4  17.0          8           302        140    3449          10.5     70   

   origin                        car  Unnamed: 9  Unnamed: 10  Unnamed: 11  \
0       1  chevrolet chevelle malibu         NaN          NaN          NaN   
1       1          buick skylark 320         NaN          NaN          NaN   
2       1         plymouth satellite         NaN          NaN          NaN   
3       1              amc rebel sst         NaN          NaN          NaN   
4       1                ford torino         NaN          NaN          NaN   

   Unnamed: 12  Unnamed: 13  
0          NaN          NaN  
1          NaN          NaN  
2          NaN          NaN  
3          NaN          NaN  
4          NaN          NaN

In [42]:
cars1.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 198 entries, 0 to 197
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           198 non-null    float64
 1   cylinders     198 non-null    int64  
 2   displacement  198 non-null    int64  
 3   horsepower    198 non-null    object 
 4   weight        198 non-null    int64  
 5   acceleration  198 non-null    float64
 6   model         198 non-null    int64  
 7   origin        198 non-null    int64  
 8   car           198 non-null    object 
 9   Unnamed: 9    0 non-null      float64
 10  Unnamed: 10   0 non-null      float64
 11  Unnamed: 11   0 non-null      float64
 12  Unnamed: 12   0 non-null      float64
 13  Unnamed: 13   0 non-null      float64
dtypes: float64(7), int64(5), object(2)
memory usage: 21.8+ KB


In [44]:
cars2.head()



mpg  cylinders  displacement horsepower  weight  acceleration  model  \
0  33.0          4            91         53    1795          17.4     76   
1  20.0          6           225        100    3651          17.7     76   
2  18.0          6           250         78    3574          21.0     76   
3  18.5          6           250        110    3645          16.2     76   
4  17.5          6           258         95    3193          17.8     76   

   origin                 car  
0       3         honda civic  
1       1      dodge aspen se  
2       1   ford granada ghia  
3       1  pontiac ventura sj  
4       1       amc pacer d/l

In [45]:
cars2.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           200 non-null    float64
 1   cylinders     200 non-null    int64  
 2   displacement  200 non-null    int64  
 3   horsepower    200 non-null    object 
 4   weight        200 non-null    int64  
 5   acceleration  200 non-null    float64
 6   model         200 non-null    int64  
 7   origin        200 non-null    int64  
 8   car           200 non-null    object 
dtypes: float64(2), int64(5), object(2)
memory usage: 14.2+ KB


### Step 4. Note that the first DataFrame cars1 has some unnamed blank columns. Fix it by removing all the unamed columns from the DataFrame permanently. Then show the first 5 rows of the new DataFrame cars1.

In [47]:
cars1.dropna(axis = 1, inplace = True)
cars1.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
0  18.0          8           307        130    3504          12.0     70   
1  15.0          8           350        165    3693          11.5     70   
2  18.0          8           318        150    3436          11.0     70   
3  16.0          8           304        150    3433          12.0     70   
4  17.0          8           302        140    3449          10.5     70   

   origin                        car  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

### Step 5. Now check whether cars1 and cars2 contains the exact same column names. You may want to use comparison operations.

In [48]:
filter = (cars1.columns == cars2.columns)
filter




array([ True,  True,  True,  True,  True,  True,  True,  True,  True])

### Step 6. Glue the two DataFrames cars1 and cars2 together into a single DataFrame called `cars` by using concatenation. Ensure the change in inplace. 

In [49]:
cars = pd.concat(([cars1, cars2]))


cars

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
0    18.0          8           307        130    3504          12.0     70   
1    15.0          8           350        165    3693          11.5     70   
2    18.0          8           318        150    3436          11.0     70   
3    16.0          8           304        150    3433          12.0     70   
4    17.0          8           302        140    3449          10.5     70   
..    ...        ...           ...        ...     ...           ...    ...   
195  27.0          4           140         86    2790          15.6     82   
196  44.0          4            97         52    2130          24.6     82   
197  32.0          4           135         84    2295          11.6     82   
198  28.0          4           120         79    2625          18.6     82   
199  31.0          4           119         82    2720          19.4     82   

     origin                        car  
0         1  chevrolet chevelle malibu  
1         1          buick skylark 320  
2         1         plymouth satellite  
3         1              amc rebel sst  
4         1                ford torino  
..      ...                        ...  
195       1            ford mustang gl  
196       2                  vw pickup  
197       1              dodge rampage  
198       1                ford ranger  
199       1                 chevy s-10  

[398 rows x 9 columns]

### Step 7. Create a new column named `owners` to the DataFrame `cars`, that contains a random integer number from in the range of 10,000 to 20,000 (inclusive) to represent the ID of the owner. Demonstrate your code works by showing the last 5 rows of the resulting DataFrame.

In [50]:
cars['owners'] = np.random.randint(10000, 20000, cars.shape[0])

cars.tail()



mpg  cylinders  displacement horsepower  weight  acceleration  model  \
195  27.0          4           140         86    2790          15.6     82   
196  44.0          4            97         52    2130          24.6     82   
197  32.0          4           135         84    2295          11.6     82   
198  28.0          4           120         79    2625          18.6     82   
199  31.0          4           119         82    2720          19.4     82   

     origin              car  owners  
195       1  ford mustang gl   16454  
196       2        vw pickup   18548  
197       1    dodge rampage   13151  
198       1      ford ranger   15988  
199       1       chevy s-10   12247

### Step 8. Did you notice that your new DataFrame cars contains 398 rows, however, the last 5 rows you just printed in step 7 are with index "195, 196, 197, 198, 199". Shouldn't they be "393, 394, 395, 396, 397"? Explain the reason. 

Your answer: With the conactnation the index of cars2 starts back from 0

### Now fix the index issue here and show the resulting DataFrame. 

In [55]:
cars = cars.reset_index()



cars

index   mpg  cylinders  displacement horsepower  weight  acceleration  \
0        0  18.0          8           307        130    3504          12.0   
1        1  15.0          8           350        165    3693          11.5   
2        2  18.0          8           318        150    3436          11.0   
3        3  16.0          8           304        150    3433          12.0   
4        4  17.0          8           302        140    3449          10.5   
..     ...   ...        ...           ...        ...     ...           ...   
393    195  27.0          4           140         86    2790          15.6   
394    196  44.0          4            97         52    2130          24.6   
395    197  32.0          4           135         84    2295          11.6   
396    198  28.0          4           120         79    2625          18.6   
397    199  31.0          4           119         82    2720          19.4   

     model  origin                        car  owners  
0       70       1  chevrolet chevelle malibu   12149  
1       70       1          buick skylark 320   12855  
2       70       1         plymouth satellite   10052  
3       70       1              amc rebel sst   12621  
4       70       1                ford torino   12815  
..     ...     ...                        ...     ...  
393     82       1            ford mustang gl   16454  
394     82       2                  vw pickup   18548  
395     82       1              dodge rampage   13151  
396     82       1                ford ranger   15988  
397     82       1                 chevy s-10   12247  

[398 rows x 11 columns]

### Step 9. Find out which car model on column `car` has the largest amount of owners. 

In [52]:
cars.groupby(by = 'car')['owners'].sum().sort_values(ascending = False).head(n=1).idxmax()



'ford pinto'

### Step 10. What are the top 5 car models that have the highest mpg winthin the weight range of 2000 and 3000?

In [57]:
filter = (cars['weight'] > 2000) & (cars['weight'] <3000)
new_cars = cars[filter]
new_cars.sort_values(by = 'mpg', ascending = False).head()['car']


322               mazda glc
325    vw rabbit c (diesel)
394               vw pickup
326      vw dasher (diesel)
309               vw rabbit
Name: car, dtype: object

# Great job!